File: app/src/main/java/suzp1984/github/io/exapidemo/graphics/PhagocyteView.java
package suzp1984.github.io.exapidemo.graphics;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

/**
 * Created by jacobsu on 9/29/16.
 */

public class PhagocyteView extends View {

    private Paint mPaint;
    private Bitmap mBitmap;

    public PhagocyteView(Context context) {
        super(context);

        init();
    }

    public PhagocyteView(Context context, AttributeSet attrs) {
        super(context, attrs);

        init();
    }

    public PhagocyteView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        init();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {

        mBitmap = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int minW = getPaddingLeft() + getPaddingRight() + getSuggestedMinimumWidth();
        int w = resolveSizeAndState(minW, widthMeasureSpec, 1);

        int minH = getPaddingBottom() + getPaddingTop() + getSuggestedMinimumHeight();
        int h = resolveSizeAndState(minH, heightMeasureSpec, 1);

        setMeasuredDimension(w, h);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        int w = canvas.getWidth();
        int h = canvas.getHeight();

        for (int j = 0; j < h; j++)
            for (int i = 0; i < w; i++) {
                double s = 3.0/(double)(j+99);
                double y = (j + Math.sin((i*i + Math.pow(j - 700, 2.0)*5.0)/100.0/w)*35.0) * s;

                int r = ((int)((i + w)*s + y)%2 + (int)((w*2 - i)*s + y)%2)*127;
                int g = ((int)(5*((i + w)*s+y))%2 + (int)(5*((w*2 - i)*s + y))%2)*127;
                int b = ((int)(29*((i + w)*s+y))%2 + (int)(29*((w*2 - i)*s + y))%2)*127;

                mBitmap.setPixel(i, j, Color.rgb(r, g, b));
            }

        canvas.drawBitmap(mBitmap, 0, 0, mPaint);
    }

    private void init() {
        mPaint = new Paint();
    }
}

File: app/src/main/kotlin/suzp1984/github/io/exapidemo/graphics/PhagocyteView.kt
package suzp1984.github.io.exapidemo.graphics

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View

/**
 * Created by jacobsu on 9/29/16.
 */

class PhagocyteView : View {

    private lateinit var mPaint: Paint
    private var mBitmap: Bitmap? = null

    constructor(context: Context) : super(context) {

        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {

        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {

        init()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {

        mBitmap = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val minW = paddingLeft + paddingRight + suggestedMinimumWidth
        val w = View.resolveSizeAndState(minW, widthMeasureSpec, 1)

        val minH = paddingBottom + paddingTop + suggestedMinimumHeight
        val h = View.resolveSizeAndState(minH, heightMeasureSpec, 1)

        setMeasuredDimension(w, h)
    }

    override fun onDraw(canvas: Canvas) {
        if (mBitmap == null) {
            return
        }

        val w = canvas.width
        val h = canvas.height

        for (j in 0..h - 1)
            for (i in 0..w - 1) {
                val s = 3.0 / (j + 99).toDouble()
                val y = (j + Math.sin((i * i + Math.pow((j - 700).toDouble(), 2.0) * 5.0) / 100.0 / w.toDouble()) * 35.0) * s

                val r = (((i + w) * s + y).toInt() % 2 + ((w * 2 - i) * s + y).toInt() % 2) * 127
                val g = ((5 * ((i + w) * s + y)).toInt() % 2 + (5 * ((w * 2 - i) * s + y)).toInt() % 2) * 127
                val b = ((29 * ((i + w) * s + y)).toInt() % 2 + (29 * ((w * 2 - i) * s + y)).toInt() % 2) * 127

                mBitmap?.setPixel(i, j, Color.rgb(r, g, b))
            }

        canvas.drawBitmap(mBitmap, 0f, 0f, mPaint)
    }

    private fun init() {
        mPaint = Paint()
    }
}
